src/xwrap.rs

```diff
@@ -2,340 +2,248 @@
+            format,
+            visual,
+            byte_order,
+            data: img.data,
+        })
     }
 
-    pub fn get_screen_rects(&self, root: xlib::Window) -> Option<ScreenRectIter<'_>> {
-        unsafe {
-            let xrr_res = xrandr::XRRGetScreenResourcesCurrent(self.handle, root);
-
-            if xrr_res.is_null() {
-                None
-            } else {
-                Some(ScreenRectIter {
-                    dpy: self,
-                    res: xrr_res,
-                    crtcs: slice::from_raw_parts((*xrr_res).crtcs, (*xrr_res).ncrtc as usize),
-                    i: 0,
-                })
-            }
+    pub fn get_screen_rects(&self) -> Option<Vec<util::Rect>> {
+        let cookie = self
+            .conn
+            .randr_get_screen_resources_current(self.root())
+            .ok()?;
+        let res = cookie.reply().ok()?;
+
+        let mut rects = Vec::new();
+        for crtc in res.crtcs {
+            let cookie = self
+                .conn
+                .randr_get_crtc_info(crtc, res.config_timestamp)
+                .ok()?;
+            let info = cookie.reply().ok()?;
+            rects.push(util::Rect {
+                x: info.x as i32,
+                y: info.y as i32,
+                w: info.width as i32,
+                h: info.height as i32,
+            });
         }
+        Some(rects)
     }
 
-    pub fn get_cursor_position(&self, window: xlib::Window) -> Option<util::Point> {
-        let mut x = 0;
-        let mut y = 0;
+    pub fn get_cursor_position(&self) -> Option<util::Point> {
+        let cookie = self.conn.query_pointer(self.root()).ok()?;
+        let pointer = cookie.reply().ok()?;
 
-        unsafe {
-            if xlib::XQueryPointer(
-                self.handle,
-                window,
-                &mut 0,
-                &mut 0,
-                &mut x,
-                &mut y,
-                &mut 0,
-                &mut 0,
-                &mut 0,
-            ) == 0
-            {
-                None
-            } else {
-                Some(util::Point { x, y })
-            }
-        }
-    }
-}
-
-impl Drop for Display {
-    fn drop(&mut self) {
-        unsafe {
-            xlib::XCloseDisplay(self.handle);
-        }
+        Some(util::Point {
+            x: pointer.win_x as i32,
+            y: pointer.win_y as i32,
+        })
     }
 }
 
 impl Image {
-    fn from_raw_ximage(ximage: *mut xlib::XImage) -> Option<Image> {
-        if ximage.is_null() {
-            None
-        } else {
-            Some(Image { handle: ximage })
-        }
-    }
-
     pub fn to_image_buffer(&self) -> Option<RgbaImage> {
-        let img = unsafe { &*self.handle };
-
-        if (img.red_mask, img.green_mask, img.blue_mask) == (0xF800, 0x07E0, 0x001F) {
+        if (
+            self.visual.red_mask,
+            self.visual.green_mask,
+            self.visual.blue_mask,
+        ) == (0xF800, 0x07E0, 0x001F)
```

Comment:
Might be worth to extract into a `const`.

Reply:
This is kind of the same thing as the `all_planes` situation. These values only get used here. They do look a bit like a magic number, but at least subjectively I think it can be inferred from context what this is.